src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Measurement.cs

```diff
@@ -0,0 +1,118 @@
+        public Measurement(T value)
+        {
+            _tags = Instrument.EmptyTags;
+            Value = value;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the Measurement using the value and the list of tags.
+        /// </summary>
+        /// <param name="value">The measurement value.</param>
+        /// <param name="tags">The measurement associated tags list.</param>
+        public Measurement(T value, IEnumerable<KeyValuePair<string, object?>>? tags)
+        {
+            _tags = ToArray(tags);
+            Value = value;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the Measurement using the value and the list of tags.
+        /// </summary>
+        /// <param name="value">The measurement value.</param>
+        /// <param name="tags">The measurement associated tags list.</param>
+        public Measurement(T value, params KeyValuePair<string, object?>[]? tags)
+        {
+            if (tags is not null)
+            {
+                _tags = new KeyValuePair<string, object?>[tags.Length];
+                tags.CopyTo(_tags, 0);
+            }
+            else
+            {
+                _tags = Instrument.EmptyTags;
+            }
+
+            Value = value;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the Measurement using the value and the list of tags.
+        /// </summary>
+        /// <param name="value">The measurement value.</param>
+        /// <param name="tags">The measurement associated tags list.</param>
+        public Measurement(T value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
+        {
+            _tags = tags.ToArray();
+            Value = value;
+        }
+
+        /// <summary>
+        /// Gets the measurement tags list.
+        /// </summary>
+        public ReadOnlySpan<KeyValuePair<string, object?>> Tags => _tags.AsSpan();
+
+        /// <summary>
+        /// Gets the measurement value.
+        /// </summary>
+        public T Value { get; }
+
+        // Private helper to copy IEnumerable to array. We have it to avoid adding dependencies on System.Linq
+        private static KeyValuePair<string, object?>[] ToArray(IEnumerable<KeyValuePair<string, object?>>? tags)
+        {
+            if (tags is null)
+            {
+                return Instrument.EmptyTags;
+            }
+
+            int count = 0;
+            using (IEnumerator<KeyValuePair<string, object?>> e = tags.GetEnumerator())
+            {
+                checked
+                {
+                    while (e.MoveNext())
+                    {
+                        count++;
+                    }
+                }
+            }
+
+            int index = 0;
+            KeyValuePair<string, object?>[] tagsArray = new KeyValuePair<string, object?>[count];
+            using (IEnumerator<KeyValuePair<string, object?>> e = tags.GetEnumerator())
```

Comment:
Enumerating an input enumerable twice like this is generally frowned upon... you have no idea how expensive iterating through it is, what side effects it might have, etc.